### PR TITLE
JVM_IR: fix a couple issues with inlining JVM assertions

### PR DESCRIPTION
**Pull request by pyos**

### CODE REVIEW

**Comment by pyos** [comment]
@max-kammerer ping?

File: compiler/testData/diagnostics/foreignAnnotationsTests/tests/rxjava3Default.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package
```
Comment (max-kammerer) [inline compiler/testData/diagnostics/foreignAnnotationsTests/tests/rxjava3Default.txt:1]: Are rx* tests added to pr by mistake?

Comment (pyos) [reply compiler/testData/diagnostics/foreignAnnotationsTests/tests/rxjava3Default.txt]: Yes, looks like these files were autogenerated when I ran `:compiler:tests-common-new:test`. Someone should probably commit them.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt [hunk 2]
```diff
@@ -197,22 +197,19 @@ class ClassCodegen private constructor(
...
+            IrSetFieldImpl(startOffset, endOffset, symbol, null, initializer!!.expression, context.irBuiltIns.unitType)
+        }
         if (generatingClInit) {
-            // Too late to modify the IR; have to ask the currently active `ExpressionCodegen`
-            // to generate this statement directly.
+            // Too late to modify the IR; have to ask the currently active `ExpressionCodegen` to generate this statement
+            // directly. At least we know that nothing before this point uses the field.
             return init
         }
         val classInitializer = irClass.functions.singleOrNull { it.name.asString() == "<clinit>" } ?: irClass.addFunction {
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt:215]: What will do of clinit already generated (or it's last operation)? // cc @ilmirus

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt [hunk 3]
```diff
@@ -558,21 +558,17 @@ class MethodInliner(
...
-                                } else {
-                                    assert(inliningContext.parent != null) {
-                                        "$inliningContext parent shall not be null"
-                                    }
-                                    inliningContext.parent!!.generateAssertField = true
-                                }
-                            } else {
-                                inliningContext.generateAssertField = true
-                            }
+                            when {
```
Comment (max-kammerer) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt:561]: Are all these cases are checked by added test?

**Comment by max-kammerer** [comment]
@ilmirus please also take a look

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt (hunk 2 above)
Comment (pyos) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt:215]: `ClassCodegen.generate` specifically delays generating clinit until after all methods for this reason

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt (hunk 3 above)
Comment (pyos) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt:561]: Yes

**Review by ilmirus — APPROVED** [review]

**Comment by max-kammerer** [comment]
@pyos  Thank you!

**Pull request merged**